src/google/protobuf/compiler/csharp/csharp_source_generator_base.cc

```diff
@@ -59,7 +59,10 @@ void SourceGeneratorBase::WriteGeneratedCodeAttributes(io::Printer* printer) {
 }
 
 std::string SourceGeneratorBase::class_access_level() {
-  return IsDescriptorProto(descriptor_) ? "internal" : "public";  // public_classes is always on.
+  if(IsDescriptorProto(descriptor_)) {
+    return "internal";
+  }
+  return this->options()->internal_access ? "internal": "public";
```

Comment:
We could either remove the `IsDescriptorProto(descriptor_)` call here and just change the generate_protos.sh file, or we could use

```
return IsDescriptorProto(descriptor_) || this->options()->internal_access ? "internal" : "public";
```

I don't mind either, particularly - but I think it's clearer to have one joint condition than an early return for descriptor protos and then a conditional.